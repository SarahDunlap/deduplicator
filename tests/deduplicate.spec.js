const deduplicate = require('../lib/deduplicate');
const { ERRORS } = require('../lib/resources');
const { getExpectedChangeValues, getMockLead, getMockLeads } = require('./utils');

describe('Deduplicate', () => {
    describe('Results', () => {
        test('if one lead, should match snapshot and have no changes detected message', async () => {
            const data = deduplicate(getMockLead());
            expect(data.logContent.search('No changes detected')).toBeGreaterThan(-1);
            expect(data.logContent).toMatchSnapshot();
        });

        test('if two duplicate leads, should match snapshot and have exact duplicate message', async () => {
            // get mock lead, push lead object to lead array
            const mockLead = getMockLead();
            const onlyLead = mockLead.leads[0];
            mockLead.leads.push(onlyLead);

            const { _id: leadId } = mockLead.leads[0];

            const data = deduplicate(mockLead);

            expect(data.logContent.search('No changes detected')).toBe(-1);
            expect(data.logContent.search(`Lead Id: ${leadId} has an exact duplicate`)).toBeGreaterThan(-1);
            expect(data).toMatchSnapshot();
        });

        test('if two leads with same id, should match snapshot', async () => {
            const mockLeads = getMockLeads();
            const { _id: firstId } = mockLeads.leads[0];
            mockLeads.leads[1]._id = firstId;
     
            const data = deduplicate(mockLeads);
            expect(data).toMatchSnapshot();
        });

        test('if two leads with same email, should match snapshot', async () => {
            const mockLeads = getMockLeads();

            // set email
            const { email: firstEmail } = mockLeads.leads[0];
            mockLeads.leads[1].email = firstEmail;

            const data = deduplicate(mockLeads);
            expect(data).toMatchSnapshot();
        });

        test('if two leads with same email and same date, should match snapshot', async () => {
            const mockLeads = getMockLeads();

            const { _id: firstId, email: firstEmail, entryDate: firstDate } = mockLeads.leads[0];
            const { _id: secondId } = mockLeads.leads[1];

            // setup second lead
            mockLeads.leads[1] = {
                ...mockLeads.leads[1],
                email: firstEmail,
                entryDate: firstDate
            }
            
            const data = deduplicate(mockLeads);
            expect(data).toMatchSnapshot();

            // check order of change history (firstId ---> secondId)
            const expectedText = getExpectedChangeValues(firstId, secondId);
            expect(data.logContent.search(expectedText)).toBeGreaterThan(-1);
        });

        test('if two leads with same email and first lead has later date, should match snapshot', async () => {
            const mockLeads = getMockLeads();
            const { _id: firstId, email: firstEmail } = mockLeads.leads[0];
            const { _id: secondId, entryDate: secondDate } = mockLeads.leads[1];

            // change date
            const firstFourChar = secondDate.slice(0,4);
            const newYear = (parseInt(firstFourChar) + 4).toString();
            const newDate = newYear + secondDate.slice(4);
            mockLeads.leads[0].entryDate = newDate;

            // change email
            mockLeads.leads[1].email = firstEmail;
            
            const data = deduplicate(mockLeads);
            expect(data).toMatchSnapshot();

            // check order of change history (secondId ---> firstId)
            const expectedText = getExpectedChangeValues(secondId, firstId);
            expect(data.logContent.search(expectedText)).toBeGreaterThan(-1);
        });

        test('if two leads with same email but different casing, should keep one and match snapshot', async () => {
            const mockLeads = getMockLeads();

            const { _id: firstId, email: firstEmail } = mockLeads.leads[0];
            const { _id: secondId } = mockLeads.leads[1];

            // set email
            mockLeads.leads[1].email = firstEmail.toUpperCase();

            const data = deduplicate(mockLeads);

            // check order (firstId ----> secondId)
            const expectedText = getExpectedChangeValues(firstId, secondId);
            expect(data.logContent.search(expectedText)).toBeGreaterThan(-1);

            expect(data).toMatchSnapshot();
        });
    });

    describe('Errors', () => {
        test('if no leads, should throw error', async () => {
            expect(() => deduplicate({leads: []})).toThrow(ERRORS.NO_LEADS);
        });

        test('if lead missing id, email, or date, should throw error', async () => {
            // id
            let mockLeads = getMockLeads();
            delete mockLeads.leads[0]._id
            expect(() => deduplicate(mockLeads)).toThrow(ERRORS.MISSING_DATA_IN_LEAD);

            // email
            mockLeads = getMockLeads();
            delete mockLeads.leads[0].email
            expect(() => deduplicate(mockLeads)).toThrow(ERRORS.MISSING_DATA_IN_LEAD);

            // date
            mockLeads = getMockLeads();
            delete mockLeads.leads[1].entryDate
            expect(() => deduplicate(mockLeads)).toThrow(ERRORS.MISSING_DATA_IN_LEAD);
        });
    });
})