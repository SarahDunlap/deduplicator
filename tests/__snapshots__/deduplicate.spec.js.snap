// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deduplicate Results if one lead, should match snapshot and have no changes detected message 1`] = `
"
Original Lead List

{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        }
    ]
}

------------------------------

Unique Lead List

[
    {
        \\"_id\\": \\"jklw\\",
        \\"email\\": \\"test@test.com\\",
        \\"firstName\\": \\"Peach\\",
        \\"lastName\\": \\"Smith\\",
        \\"address\\": \\"123 Main St\\",
        \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
    }
]

------------------------------

Change History

No changes detected.
"
`;

exports[`Deduplicate Results if two duplicate leads, should match snapshot and have exact duplicate message 1`] = `
Object {
  "logContent": "
Original Lead List

{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        },
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        }
    ]
}

------------------------------

Unique Lead List

[
    {
        \\"_id\\": \\"jklw\\",
        \\"email\\": \\"test@test.com\\",
        \\"firstName\\": \\"Peach\\",
        \\"lastName\\": \\"Smith\\",
        \\"address\\": \\"123 Main St\\",
        \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
    }
]

------------------------------

Change History

Duplicate entry detected for email field (test@test.com).
Lead Id: jklw has an exact duplicate.
",
  "uniqueLeadsContent": "{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        }
    ]
}",
}
`;

exports[`Deduplicate Results if two leads with same email and first lead has later date, should match snapshot 1`] = `
Object {
  "logContent": "
Original Lead List

{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2018-05-07T18:30:20+00:00\\"
        },
        {
            \\"_id\\": \\"esdf\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Gary\\",
            \\"lastName\\": \\"Parker\\",
            \\"address\\": \\"432 Main St\\",
            \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
        }
    ]
}

------------------------------

Unique Lead List

[
    {
        \\"_id\\": \\"jklw\\",
        \\"email\\": \\"test@test.com\\",
        \\"firstName\\": \\"Peach\\",
        \\"lastName\\": \\"Smith\\",
        \\"address\\": \\"123 Main St\\",
        \\"entryDate\\": \\"2018-05-07T18:30:20+00:00\\"
    }
]

------------------------------

Change History

Duplicate entry detected for email field (test@test.com).
_id: esdf ----> jklw
firstName: Gary ----> Peach
lastName: Parker ----> Smith
address: 432 Main St ----> 123 Main St
entryDate: 2014-05-07T18:30:20+00:00 ----> 2018-05-07T18:30:20+00:00

",
  "uniqueLeadsContent": "{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2018-05-07T18:30:20+00:00\\"
        }
    ]
}",
}
`;

exports[`Deduplicate Results if two leads with same email and same date, should match snapshot 1`] = `
Object {
  "logContent": "
Original Lead List

{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        },
        {
            \\"_id\\": \\"esdf\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Gary\\",
            \\"lastName\\": \\"Parker\\",
            \\"address\\": \\"432 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        }
    ]
}

------------------------------

Unique Lead List

[
    {
        \\"_id\\": \\"esdf\\",
        \\"email\\": \\"test@test.com\\",
        \\"firstName\\": \\"Gary\\",
        \\"lastName\\": \\"Parker\\",
        \\"address\\": \\"432 Main St\\",
        \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
    }
]

------------------------------

Change History

Duplicate entry detected for email field (test@test.com).
_id: jklw ----> esdf
firstName: Peach ----> Gary
lastName: Smith ----> Parker
address: 123 Main St ----> 432 Main St

",
  "uniqueLeadsContent": "{
    \\"leads\\": [
        {
            \\"_id\\": \\"esdf\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Gary\\",
            \\"lastName\\": \\"Parker\\",
            \\"address\\": \\"432 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        }
    ]
}",
}
`;

exports[`Deduplicate Results if two leads with same email but different casing, should keep one and match snapshot 1`] = `
Object {
  "logContent": "
Original Lead List

{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        },
        {
            \\"_id\\": \\"esdf\\",
            \\"email\\": \\"TEST@TEST.COM\\",
            \\"firstName\\": \\"Gary\\",
            \\"lastName\\": \\"Parker\\",
            \\"address\\": \\"432 Main St\\",
            \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
        }
    ]
}

------------------------------

Unique Lead List

[
    {
        \\"_id\\": \\"esdf\\",
        \\"email\\": \\"TEST@TEST.COM\\",
        \\"firstName\\": \\"Gary\\",
        \\"lastName\\": \\"Parker\\",
        \\"address\\": \\"432 Main St\\",
        \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
    }
]

------------------------------

Change History

Duplicate entry detected for email field (test@test.com).
_id: jklw ----> esdf
email: test@test.com ----> TEST@TEST.COM
firstName: Peach ----> Gary
lastName: Smith ----> Parker
address: 123 Main St ----> 432 Main St
entryDate: 2014-05-07T17:30:20+00:00 ----> 2014-05-07T18:30:20+00:00

",
  "uniqueLeadsContent": "{
    \\"leads\\": [
        {
            \\"_id\\": \\"esdf\\",
            \\"email\\": \\"TEST@TEST.COM\\",
            \\"firstName\\": \\"Gary\\",
            \\"lastName\\": \\"Parker\\",
            \\"address\\": \\"432 Main St\\",
            \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
        }
    ]
}",
}
`;

exports[`Deduplicate Results if two leads with same email, should match snapshot 1`] = `
Object {
  "logContent": "
Original Lead List

{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        },
        {
            \\"_id\\": \\"esdf\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Gary\\",
            \\"lastName\\": \\"Parker\\",
            \\"address\\": \\"432 Main St\\",
            \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
        }
    ]
}

------------------------------

Unique Lead List

[
    {
        \\"_id\\": \\"esdf\\",
        \\"email\\": \\"test@test.com\\",
        \\"firstName\\": \\"Gary\\",
        \\"lastName\\": \\"Parker\\",
        \\"address\\": \\"432 Main St\\",
        \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
    }
]

------------------------------

Change History

Duplicate entry detected for email field (test@test.com).
_id: jklw ----> esdf
firstName: Peach ----> Gary
lastName: Smith ----> Parker
address: 123 Main St ----> 432 Main St
entryDate: 2014-05-07T17:30:20+00:00 ----> 2014-05-07T18:30:20+00:00

",
  "uniqueLeadsContent": "{
    \\"leads\\": [
        {
            \\"_id\\": \\"esdf\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Gary\\",
            \\"lastName\\": \\"Parker\\",
            \\"address\\": \\"432 Main St\\",
            \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
        }
    ]
}",
}
`;

exports[`Deduplicate Results if two leads with same id, should match snapshot 1`] = `
Object {
  "logContent": "
Original Lead List

{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"test@test.com\\",
            \\"firstName\\": \\"Peach\\",
            \\"lastName\\": \\"Smith\\",
            \\"address\\": \\"123 Main St\\",
            \\"entryDate\\": \\"2014-05-07T17:30:20+00:00\\"
        },
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"what@test.com\\",
            \\"firstName\\": \\"Gary\\",
            \\"lastName\\": \\"Parker\\",
            \\"address\\": \\"432 Main St\\",
            \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
        }
    ]
}

------------------------------

Unique Lead List

[
    {
        \\"_id\\": \\"jklw\\",
        \\"email\\": \\"what@test.com\\",
        \\"firstName\\": \\"Gary\\",
        \\"lastName\\": \\"Parker\\",
        \\"address\\": \\"432 Main St\\",
        \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
    }
]

------------------------------

Change History

Duplicate entry detected for _id field (jklw).
email: test@test.com ----> what@test.com
firstName: Peach ----> Gary
lastName: Smith ----> Parker
address: 123 Main St ----> 432 Main St
entryDate: 2014-05-07T17:30:20+00:00 ----> 2014-05-07T18:30:20+00:00

",
  "uniqueLeadsContent": "{
    \\"leads\\": [
        {
            \\"_id\\": \\"jklw\\",
            \\"email\\": \\"what@test.com\\",
            \\"firstName\\": \\"Gary\\",
            \\"lastName\\": \\"Parker\\",
            \\"address\\": \\"432 Main St\\",
            \\"entryDate\\": \\"2014-05-07T18:30:20+00:00\\"
        }
    ]
}",
}
`;
